File: src/Starfield2026.BgEditor/backend/src/lib/ohana/Models/PocketMonsters/CM.ts
/**
 * Loads a CM overworld character model from Pokemon.
 * Ported from OhanaCli.Formats.Models.PocketMonsters.CM (C#).
 */

import { BinaryReader } from '../../Core/BinaryReader.js';
import { PkmnContainer } from '../../Containers/PkmnContainer.js';
import { GfModel } from './GfModel.js';
import { GfMotion } from './GfMotion.js';
import { OModelGroup } from '../../Core/RenderBase.js';

export class CM {
  static load(data: BinaryReader): OModelGroup {
    const container = PkmnContainer.load(data);
    const models = GfModel.load(BinaryReader.fromBuffer(container.content[0].data!));

    if (container.content.length > 1 && container.content[1].data) {
      try {
        const anms = GfMotion.load(BinaryReader.fromBuffer(container.content[1].data));
        for (const anm of anms) models.skeletalAnimation.list.push(anm);
      } catch {
        // Animation data may not be present or valid in all CM containers
      }
    }

    return models;
  }
}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Garc/Models/PocketMonsters/CM.cs
using MiniToolbox.Garc.Containers;
using System;
using System.Collections.Generic;
using System.IO;

namespace MiniToolbox.Garc.Models.PocketMonsters
{
    class CM
    {
        /// <summary>
        ///     Loads a CM overworld character model from Pokémon.
        /// </summary>
        /// <param name="data">The data</param>
        /// <returns>The Model group with the character meshes</returns>
        public static RenderBase.OModelGroup load(Stream data)
        {
            RenderBase.OModelGroup models = new RenderBase.OModelGroup();

            OContainer container = PkmnContainer.load(data);
            models = GfModel.load(new MemoryStream(container.content[0].data));

            if (container.content.Count > 1 && container.content[1].data != null && container.content[1].data.Length > 4)
            {
                try
                {
                    List<RenderBase.OSkeletalAnimation> anms = GfMotion.load(new MemoryStream(container.content[1].data));
                    foreach (RenderBase.OSkeletalAnimation anm in anms)
                    {
                        models.skeletalAnimation.list.Add(anm);
                    }
                }
                catch (Exception)
                {
                    // Animation data may be malformed for some entries
                }
            }

            return models;
        }
    }
}
